File: vercel-deployment/src/lib/encryption.ts
/**
 * Encryption utilities for API key storage (TypeScript version)
 */

let encryptionKey: CryptoKey | null = null;

async function getEncryptionKey(): Promise<CryptoKey> {
  if (encryptionKey) return encryptionKey;

  const keyB64 = process.env.ENCRYPTION_KEY;
  if (!keyB64) {
    throw new Error("ENCRYPTION_KEY not set in environment");
  }

  // Decode base64 key
  const keyBytes = Uint8Array.from(atob(keyB64), c => c.charCodeAt(0));
  
  // Import as AES-GCM key
  encryptionKey = await crypto.subtle.importKey(
    "raw",
    keyBytes,
    { name: "AES-GCM" },
    false,
    ["encrypt", "decrypt"]
  );
  
  return encryptionKey;
}

export async function encryptApiKey(apiKey: string): Promise<string> {
  const key = await getEncryptionKey();
  const encoder = new TextEncoder();
  const data = encoder.encode(apiKey);
  
  // Generate random IV
  const iv = crypto.getRandomValues(new Uint8Array(12));
  
  // Encrypt
  const encrypted = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv },
    key,
    data
  );
  
  // Combine IV + encrypted data
  const result = new Uint8Array(iv.length + encrypted.byteLength);
  result.set(iv);
  result.set(new Uint8Array(encrypted), iv.length);
  
  // Return as base64
  return btoa(String.fromCharCode.apply(null, Array.from(result)));
}

export async function decryptApiKey(encryptedKey: string): Promise<string> {
  const key = await getEncryptionKey();
  
  // Decode from base64
  const data = Uint8Array.from(atob(encryptedKey), c => c.charCodeAt(0));
  
  // Extract IV (first 12 bytes)
  const iv = data.slice(0, 12);
  const encrypted = data.slice(12);
  
  // Decrypt
  const decrypted = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv },
    key,
    encrypted
  );
  
  const decoder = new TextDecoder();
  return decoder.decode(decrypted);
}

export function generateEncryptionKey(): string {
  const array = new Uint8Array(32);
  crypto.getRandomValues(array);
  return btoa(String.fromCharCode.apply(null, Array.from(array)));
}

File: vercel-deployment/src/lib/encryption.py
"""Encryption utilities for API key storage"""
from __future__ import annotations

import base64
import os
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

_ENCRYPTION_KEY = None


def get_encryption_key() -> bytes:
    """Get or derive the encryption key from environment"""
    global _ENCRYPTION_KEY
    
    if _ENCRYPTION_KEY is not None:
        return _ENCRYPTION_KEY
    
    key_b64 = os.environ.get("ENCRYPTION_KEY")
    if not key_b64:
        # Generate a key for development (NOT for production!)
        import secrets
        key_b64 = base64.urlsafe_b64encode(secrets.token_bytes(32)).decode()
        print(f"⚠️  Generated temporary ENCRYPTION_KEY: {key_b64}")
        print("   Set ENCRYPTION_KEY in production!")
    
    try:
        _ENCRYPTION_KEY = base64.urlsafe_b64decode(key_b64)
    except Exception:
        # Derive from password if not valid base64
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=b'multillm-salt',  # In production, use per-user salt
            iterations=100000,
        )
        _ENCRYPTION_KEY = base64.urlsafe_b64encode(kdf.derive(key_b64.encode()))
    
    return _ENCRYPTION_KEY


def encrypt_api_key(api_key: str) -> str:
    """Encrypt an API key for storage"""
    key = get_encryption_key()
    f = Fernet(key)
    return f.encrypt(api_key.encode()).decode()


def decrypt_api_key(encrypted_key: str) -> str:
    """Decrypt an API key from storage"""
    key = get_encryption_key()
    f = Fernet(key)
    return f.decrypt(encrypted_key.encode()).decode()


def generate_encryption_key() -> str:
    """Generate a new encryption key (run once and store securely)"""
    import secrets
    return base64.urlsafe_b64encode(secrets.token_bytes(32)).decode()


if __name__ == "__main__":
    # Generate a key for setup
    print(f"ENCRYPTION_KEY={generate_encryption_key()}")